File: src/java/list/160.intersection-of-two-linked-lists.java
class P160 {
    // @lc app=leetcode id=160 lang=java
    // @lc code=start 
    public class Solution {
        public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
            if (headA == null || headB == null) return null;
            
            ListNode pA = headA;
            ListNode pB = headB;
            
            // Two pointer technique: when one reaches end, start from other head
            while (pA != pB) {
                pA = (pA == null) ? headB : pA.next;
                pB = (pB == null) ? headA : pB.next;
            }
            
            return pA; // Either intersection node or null
        }
    }
    // @lc code=end
}

File: src/go/list/160.intersection-of-two-linked-lists.go
package main

/*
 * @lc app=leetcode id=160 lang=golang
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}

	pA, pB := headA, headB

	for pA != pB {
		if pA == nil {
			pA = headB
		} else {
			pA = pA.Next
		}

		if pB == nil {
			pB = headA
		} else {
			pB = pB.Next
		}
	}

	return pA
}

// @lc code=end
